translated/tech/20180330-Go-On-Very-Small-Hardware-Part2.md

```diff
@@ -0,0 +1,898 @@
+# Go 最小硬件编程（第二部分）
+
+[![STM32F030F4P6](https://ziutek.github.io/images/mcu/f030-demo-board/board.jpg)](https://ziutek.github.io/2018/04/14/go_on_very_small_hardware2.html)
+
+在本文 [第一部分](https://studygolang.com/articles/22742) 的结尾，我说过要写一下关于 *接口* 的东西。我不想在这里写一篇完整或是简短的关于接口的讲稿。相反，我将会举一个简单的例子，用以说明如何定义和使用接口，同时知道如何利用通用的 *io.Writer* 接口。这里也会出现一些关于 *reflection* 和 *semihosting* 的一些名词。
+
+接口是 Go 语言的关键部分。如果你想要更多的学习它们，我建议去阅读 [Effective Go](https://golang.org/doc/effective_go.html#interfaces) 和 [Russ Cox article](https://research.swtch.com/interfaces)。
+
+## 并发闪烁 - 重温
+
+当你阅读之前那个例子的代码时，可能会注意到开启或是关闭 LED 的方式有悖于常理。*Set* 方法用于关闭 LED，同时 *Clear* 方法用于打开 LED，这是由于使用了开漏配置去驱动 LED。我们可以做什么以让代码不那么让人迷惑呢？让我们使用 *On* 和 *Off* 方法来定义 LED 类型吧。
+
+```go
+type LED struct {
+	pin gpio.Pin
+}
+
+func (led LED) On() {
+	led.pin.Clear()
+}
+
+func (led LED) Off() {
+	led.pin.Set()
+}
+```
+
+现在我们可以简单地调用 `led.On()` 和 `led.Off()` 方法了，这样就不会再引起任何疑惑。
+
+在前面所有的的例子中，我努力去使用相同的开漏配置，以让代码不会变得复杂。但是在上一个例子中，在 GND 和 PA3 引脚间连接第三个 LED，同时配置 PA3 为推拉模式，似乎更简单。接下来的例子就将会使用这种方式连接一个 LED。
+
+但是我们的新 *LED* 类型不支持推拉模式配置。实际上，我们应该叫它 *OpenDrainLED* 同时定义另外一个 *PushPullLED* 类型。
+
+```go
+type PushPullLED struct {
+	pin gpio.Pin
+}
+
+func (led PushPullLED) On() {
+	led.pin.Set()
+}
+
+func (led PushPullLED) Off() {
+	led.pin.Clear()
+}
+```
+
+注意，这两种类型都有同样的方法进行同样的工作。要是在代码上能用两种类型操作 LED，这将会是很棒的，这样就不用关心在什么时候使用的是哪一个类型了。*接口* 类型就能帮上忙了。
```

Comment:
It would be nice if the code that operates on LEDs could use both types, without paying attention to which one it uses at the moment. 
如果操作 LED 的代码可以使用两种类型，不需要关心使用的是哪一个，这将会很棒。*接口* 类型就能帮上忙了。

the code that operates on LEDs 这是从句的主语； 操作 led 的代码
从语义看，整个 if the code that operates on LEDs could use both types, without paying attention to which one it uses at the moment. 是从句